src/Microsoft.Data.SqlClient/src/Microsoft/Data/SqlClient/SqlSecurityUtility.cs

```diff
@@ -378,5 +411,50 @@ internal static void VerifyColumnMasterKeySignature(string keyStoreName, string
                 throw SQL.ColumnMasterKeySignatureVerificationFailed(keyPath);
             }
         }
+
+        internal static void ThrowIfKeyPathIsNotTrustedForServer(string serverName, string keyPath)
+        {
+            // Check against the trusted key paths
+            // Get the List corresponding to the connected server
+            if (SqlConnection.ColumnEncryptionTrustedMasterKeyPaths.TryGetValue(serverName, out IList<string> trustedKeyPaths))
+            {
+                // If the list is null or is empty or if the keyPath doesn't exist in the trusted key paths, then throw an exception.
+                if (trustedKeyPaths == null || trustedKeyPaths.Count() == 0 ||
+                    trustedKeyPaths.Any(trustedKeyPath => trustedKeyPath.Equals(keyPath, StringComparison.InvariantCultureIgnoreCase)) == false)
+                {
+                    // throw an exception since the key path is not in the trusted key paths list for this server
+                    throw SQL.UntrustedKeyPath(keyPath, serverName);
+                }
+            }
+        }
+
+        internal static bool TryGetColumnEncryptionKeyStoreProvider(string keyStoreName, out SqlColumnEncryptionKeyStoreProvider provider, SqlConnection connection, SqlCommand command)
+        {
+            Debug.Assert(!string.IsNullOrWhiteSpace(keyStoreName), "Provider name is invalid");
+
+            // Search in the system provider list.
+            if (SqlConnection.s_systemColumnEncryptionKeyStoreProviders.TryGetValue(keyStoreName, out provider))
+            {
+                return true;
+            }
+
+            // command may be null because some callers do not have a command object, eg SqlBulkCopy
+            if (command != null && command.HasColumnEncryptionKeyStoreProvidersRegistered)
+            {
+                return command.TryGetColumnEncryptionKeyStoreProvider(keyStoreName, out provider);
+            }
+
+            return connection.TryGetColumnEncryptionKeyStoreProvider(keyStoreName, out provider);
```

Comment:
The logic seems out of order.

Don't we want to check for command, then connection, then system level providers?

```suggestion
            // command may be null because some callers do not have a command object, eg SqlBulkCopy
            if (command != null && command.HasColumnEncryptionKeyStoreProvidersRegistered)
            {
                return command.TryGetColumnEncryptionKeyStoreProvider(keyStoreName, out provider);
            }

            if (connection.HasColumnEncryptionKeyStoreProvidersRegistered)
            {
                return connection.TryGetColumnEncryptionKeyStoreProvider(keyStoreName, out provider);
            }

            // Search in the system provider list.
            return SqlConnection.s_systemColumnEncryptionKeyStoreProviders.TryGetValue(keyStoreName, out provider)
```

Reply:
Yep, you're right.